synapse/handlers/sync.py

```diff
@@ -565,13 +565,12 @@ def _generate_sync_entry_for_to_device(self, sync_result_builder):
         if sync_result_builder.since_token is not None:
             since_stream_id = int(sync_result_builder.since_token.to_device_key)
 
-        if since_stream_id:
+        if since_stream_id != int(now_token.to_device_key):
             logger.debug("Deleting messages up to %d", since_stream_id)
```

Comment:
Maybe comment that this won't delete the most recent batch of messages until a new message comes along?